### PR TITLE
Honor SOURCE_DATE_EPOCH in more cases

### LINKED ISSUE #2711: Please do not put timestamps in generated files
cc @frank-trampe 
### When you open an issue for a change/improvement/feature request:
- [x] A description of the problem you're trying to solve, including _why_ you think this is a problem
- [ ] If the feature changes current behavior, reasons why your solution is better
- [x] (optional) Possible solution/fix/workaround

I'd like to propose to not put any timestamps in the generated .ttf (or any other) files. If this is not wanted generally, I'd like to propose an option which allows this.
### Reason:

The [Wine project](https://www.winehq.org/) uses a [forked fontforge](http://source.winehq.org/git/fontforge.git/) (this issue is about 1 of their 4 changes). The generated files are stored in a git repository. Recreating the files with an unpatched fontforge creates spurious diffs after every rebuild.
Generally this is related to reproducible builds (see below).
### Related things:

This has already been partly requested in issue #2389 (Generate: Don't include FFTM) by @davelab6, but afaik timestamps also land in the tables "head" (and "bhed"). I understand that while FFTM is fontforge specific, the latter two were [specified by Apple](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6head.html). I assume/hope this doesn't mean that timestamps are strictly required in order to not break 3rd party consumers.
But even in this case I can confirm that Wine works happily without timestamps.

This is also related to the **[reproducible builds effort](https://reproducible-builds.org/)**:
-  #2490 (propagate creation and modification times to TTF files, [Debian bug #774148](https://bugs.debian.org/774148)) is already fixed (95a470e941d9a20fbdaca51334e8b6b9d93cfae4).
- But there's still [Debian bug #774274](https://bugs.debian.org/774274) (fontforge: please use SOURCE_DATE_EPOCH for reproducible font modification time; was: please allow scripts to set the font modification time). This request would be obsolete if you dropped timestamps altogether.
  On the other side, [implementing SOURCE_DATE_EPOCH](https://reproducible-builds.org/specs/source-date-epoch/) would also solve the problem for Winehq maintainers if a sensible workflow is found. It's just not as trivial as for a Debian maintainer. If it helps I may file a separate issue about that. I have no patch to offer, but there are some instructions in the link.
### For reference:

We discussed the Winehq fork:
https://www.winehq.org/pipermail/wine-devel/2016-April/112893.html

The original Winehq hacks (based on 20081224):
http://source.winehq.org/git/fontforge.git/commitdiff/65974286fe1ae1700eb58e32000b13648455000f

The currently used form (rebased on v20120731-b):
https://github.com/jre-wine/fontforge/commit/23d13b4a5923745e586b0fd03f23ded6706230c3

My attempt to rebase it on 20160404 (doesn't seem to catch everything: the ttf's still differ, while they are identical with Winehq's fontforge):
https://github.com/jre-wine/fontforge/commit/c2d28ed828ab3c82ed028a22702bae056f174ba2
### Disclaimer:

I'm not using fontforge personally, neither am I a Wine developer. I'm only working on the Wine packaging in Debian. So although I really tried to get a good picture of what is still relevant, it may well be that I missed something.
Winehq and Debian are currently both using 20120731-b, but Debian is working on updating, and Winehq is willing to do the same. I tested my changes with 20160404 locally.

Thanks and greets
jre

### PR DESCRIPTION
**Pull request by gioele**
This PR is an extension of the merged PR #3152, making sure that SOURCE_DATE_EPOCH is used everywhere the current time is serialized in a file.

@volth, @bmwiedemann: could you please test this to see if further reduces or eliminate problems with reproducible build on Linux?

@volth: The MacOS issues will need to be taken care of separately.

WRT Author names, `GetAuthor` already honors the user-configurable env variable `USER`.

Closes #3134
Fixes #2711

### CODE REVIEW

File: contrib/fonttools/Makefile.am [hunk 1]
```diff
@@ -35,6 +35,8 @@ noinst_PROGRAMS = findtable pfadecrypt rmligamarks stripttc woff	\
 
 woff_LDADD = $(ZLIB_LIBS)
 dewoff_LDADD = $(ZLIB_LIBS)
+ttf2eps_LDADD = $(top_builddir)/Unicode/libgunicode.la \
```
Comment (jtanx) [inline contrib/fonttools/Makefile.am]: Seems a bit heavy to add a dependency on these dlls for just the one function (and considering the standalone nature of the rest of the contrib tools)

File: fontforge/macbinary.c [hunk 2]
```diff
@@ -1423,7 +1422,6 @@ static int DumpMacBinaryHeader(FILE *res,struct macbinaryheader *mb) {
 
 	/* Creation time, (seconds from 1/1/1904) */
     now = mactime();
-    time(&now);
```
Comment (jtanx) [inline fontforge/macbinary.c]: What's up with this?

File: fontforge/print.c [hunk 3]
```diff
@@ -1155,23 +1156,11 @@ static void dump_pdfprologue(PI *pi) {
...
     fprintf( pi->out, "  /Producer (FontForge)\n" );
-    time(&now);
-    tm = localtime(&now);
+    now = GetTime();
+    tm = gmtime(&now);
     fprintf( pi->out, "    /CreationDate (D:%04d%02d%02d%02d%02d%02d",
 	    tm->tm_year+1900, tm->tm_mon+1, tm->tm_mday, tm->tm_hour, tm->tm_min, tm->tm_sec );
-#ifdef _NO_TZSET
     fprintf( pi->out, "Z)\n" );
-#else
```
Comment (jtanx) [inline fontforge/print.c]: Why is this removed?

File: fontforge/macbinary.c (hunk 2 above)
Comment (gioele) [reply fontforge/macbinary.c]: Oops, I forgot to add the background story in the commit. :)

That `time(&now)` has been added by GW during a mass commit. If the comment is right then using `time` there is plainly wrong, as `time` counts from 1970. On the other hand, `mactime` counts from 1904, just like the comment says. In addition, in the following line the modification time is also stored and there only `mactime` is used.

I trust the comment and the overall context, thus removed that `time`. However I have put it in a separate comment, so that it will be easy to bisect should any problem arise.

File: contrib/fonttools/Makefile.am (hunk 1 above)
Comment (gioele) [reply contrib/fonttools/Makefile.am]: That dependency isn't already there only because there is a lot of code copied and pasted in `ttf2eps`. :) Same for `showttf`

We really should go though it and remove all the duplicated and outdated code. This would leave only the core ttf2eps logic there. That is the point of a `libfontforge`, isn't it? ;)

File: fontforge/print.c (hunk 3 above)
Comment (gioele) [reply fontforge/print.c]: That part of the code was there only to calculate and output the timezone offset. With no timezone there is no need to calculate it. ;)

BTW, I think it is broken on MacOS (can't remember the details, I had a brief look before realizing it could just be deleted).

**Comment by frank-trampe** [comment]
Why is it necessary to drop timezone support?

**Comment by bmwiedemann** [comment]
I think the plan is to always use UTC everywhere so that users with different timezone settings can still get the same result from fontforge calls.

**Comment by gioele** [comment]
hi @volth, hi @bmwiedemann, I added some code to make sure that the env variable `USER` is preferred to GECOS fields if `SOURCE_DATE_EPOCH`. I cannot see any other externally controlled source of non-reproducibility. Does this work for you?

I did quite a bit of testing and everything seems to work fine when `SOURCE_DATE_EPOCH` is set as well as when it is not set.

Any objections to merging this?

**Comment by volth** [comment]
I am not sure about Debian, but NixOS does not a set fake stable ```$USER``` for the build, and I anticipate the need to forge ```$USER``` may meet some objections in the community

**Comment by gioele** [comment]
What does the build system of NixOS does when an author/user _must_ be set? Does it default to something?

Or is there an equivalent `SOURCE_DATE_EPOCH` for `USER`? Something like `SOURCE_USER`? If such a thing exist we can easily support it.

**Comment by volth** [comment]
> What does the build system of NixOS does when an author/user must be set?

I guess it is something very new and fontforge-specific. Are the other build tools besides ```fontforge``` with the concept of "```author/user must be set```" ?

**Comment by gioele** [comment]
> > What does the build system of NixOS does when an author/user must be set?
>
> I guess it is something very new and fontforge-specific. Are the other build tools besides fontforge with the concept of "author/user must be set" ?

Too many. :) tar, for example, must put something in the user/group of each file it records in a tar ball. Git/SVN/Hg/CVS also require you to have an user/email set and will store it in the metadata files.

I think that the more general problem (that surely has already been dealt with by NixOS) is: "How do you deal with env variables that end up being encoded in a compiled file?".

**Comment by frank-trampe** [comment]
We can't drop timezone support without getting feedback from a lot of people. Let's either trigger UTC output when SOURCE_DATE_EPOCH is set or, if that doesn't work for some reason, set another environment variable like TIMESTAMP_TZ.

**Comment by gioele** [comment]
hi @frank-trampe, you have a point: there is a slight chance of things breaking because of the missing timezone (I can picture a script that checks against the local time using strings instead of proper time routines). Assuming sloppy code out there seems a sensible default. ;)

I rebased the branch against master, reverted most of the commit that forced UTC and added some code to conditionally use UTC if `SOURCE_DATA_EPOCH` is set.

File: contrib/fonttools/Makefile.am (hunk 1 above)
Comment (jtanx) [reply contrib/fonttools/Makefile.am]: True, but I can also see the reasoning; they have no dependencies on anything else, so they can be shipped independently of FF itself. 

For example, if I'm reading this right, the [fontforge-extras](https://packages.debian.org/sid/fontforge-extras) package ships with no dependency on fontforge itself.

**Comment by frank-trampe** [comment]
@gioele, I'd like to rebase (perhaps in a new branch and pull request) with the "Output UTC" and its revert dropped. I can do it, but it's easier for you to do since you own the branch.

**Comment by gioele** [comment]
hi @frank-trampe, I rebased the branch and squashed away the "Output UTC" commit/revert commit.

**Comment by h01ger** [comment]
the implemented logic and decissions look good to me - though I'm not familar with this code base :)

Thank you!

**Comment by gioele** [comment]
hi @frank-trampe and @jtanx, are there any outstanding objections to merging this PR?

**Comment by frank-trampe** [comment]
Go ahead!

**Comment by frank-trampe** [comment]
Or I'll do it.

File: contrib/fonttools/Makefile.am (hunk 1 above)
Comment (frank-trampe) [reply contrib/fonttools/Makefile.am]: I forgot to comment here, but our repackaging dropped fontforge-extras. I'm hoping that Debian will follow suit.

**Comment by gioele** [comment]
Well, you've been faster. Thanks :)
